### PR TITLE
Enabling RuntimeEventSource in NativeAOT

### PR DESCRIPTION
**Pull request by LakshanF**
`RuntimeEventSource` wasn't enabled in NativeAOT due to not being initialized when `EventSourceEnabled` is set, missing metadata on critical payload types [blocking reflection](https://github.com/dotnet/runtime/issues/72570) of these in `EventSource`.

Added a module initializer to Initialize `RuntimeEventSource` if `EventSourceEnabled` is set, made the required types visible to reflection, and also enabled some missing counters in `NativeAOT`.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterPayload.cs [hunk 1]
```diff
@@ -7,7 +7,12 @@
 namespace System.Diagnostics.Tracing
 {
     [EventData]
-    internal sealed class CounterPayload : IEnumerable<KeyValuePair<string, object?>>
+#if NATIVEAOT
+    public // On NativeAOT, this must be public to prevent it from getting reflection blocked.
```
[the remark this answers is not in the document]
Comment (agocke) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterPayload.cs:11]: Unfortunately I think this is stripped specifically because it's in System.Private.CoreLib and even the normal attributes don't work. @MichalStrehovsky can you link the bug for this issue?

[the remark this answers is not in the document]
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterPayload.cs:11]: #72570

[the remark this answers is not in the document]
Comment (MichalStrehovsky) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterPayload.cs:11]: Thanks for linking the issue Jan.

We already have multiple ifdefs like this in S.D.Tracing since the .NET Native days, for example:

https://github.com/dotnet/runtime/blob/e029183d484dabf02c87140415a1632a101b4647/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/PropertyValue.cs#L16-L22

We'd clean them up if/when we get rid of reflection blocking.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs [hunk 2]
```diff
@@ -108,15 +138,17 @@ protected override void OnEventCommand(EventCommandEventArgs command)
                     return gcInfo.HeapSizeBytes != 0 ? gcInfo.FragmentedBytes * 100d / gcInfo.HeapSizeBytes : 0;
                  }) { DisplayName = "GC Fragmentation", DisplayUnits = "%" };
 
-#if !NATIVEAOT // TODO
                 _committedCounter ??= new PollingCounter("gc-committed", this, () => ((double)GC.GetGCMemoryInfo().TotalCommittedBytes / 1_000_000)) { DisplayName = "GC Committed Bytes", DisplayUnits = "MB" };
+#if !NATIVEAOT // TODO
                 _exceptionCounter ??= new IncrementingPollingCounter("exception-count", this, () => Exception.GetExceptionCount()) { DisplayName = "Exception Count", DisplayRateTimeScale = new TimeSpan(0, 0, 1) };
```
Comment (MichalStrehovsky) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:143]: On NativeAOT if you want to do exception counting, you'd do it in `Exception.AppendExceptionStackFrame`.

You'd key of off isFirstFrame/isFirstRethrowFrame, depending on how it's supposed to count rethrows.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs [hunk 3]
```diff
@@ -108,15 +138,17 @@ protected override void OnEventCommand(EventCommandEventArgs command)
...
                 _gcTimeCounter ??= new PollingCounter("time-in-gc", this, () => GC.GetLastGCPercentTimeInGC()) { DisplayName = "% Time in GC since last GC", DisplayUnits = "%" };
                 _gen0SizeCounter ??= new PollingCounter("gen-0-size", this, () => GC.GetGenerationSize(0)) { DisplayName = "Gen 0 Size", DisplayUnits = "B" };
                 _gen1SizeCounter ??= new PollingCounter("gen-1-size", this, () => GC.GetGenerationSize(1)) { DisplayName = "Gen 1 Size", DisplayUnits = "B" };
                 _gen2SizeCounter ??= new PollingCounter("gen-2-size", this, () => GC.GetGenerationSize(2)) { DisplayName = "Gen 2 Size", DisplayUnits = "B" };
                 _lohSizeCounter ??= new PollingCounter("loh-size", this, () => GC.GetGenerationSize(3)) { DisplayName = "LOH Size", DisplayUnits = "B" };
                 _pohSizeCounter ??= new PollingCounter("poh-size", this, () => GC.GetGenerationSize(4)) { DisplayName = "POH (Pinned Object Heap) Size", DisplayUnits = "B" };
+#if !NATIVEAOT // TODO
                 _assemblyCounter ??= new PollingCounter("assembly-count", this, () => System.Reflection.Assembly.GetAssemblyCount()) { DisplayName = "Number of Assemblies Loaded" };
```
Comment (MichalStrehovsky) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:152]: We can take inspiration from `AppDomain.CurrentDomain.GetAssemblies()` for this. It will forever be one number though. The loaded assemblies don't change.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs [hunk 4]
```diff
@@ -108,15 +132,21 @@ protected override void OnEventCommand(EventCommandEventArgs command)
                     return gcInfo.HeapSizeBytes != 0 ? gcInfo.FragmentedBytes * 100d / gcInfo.HeapSizeBytes : 0;
                  }) { DisplayName = "GC Fragmentation", DisplayUnits = "%" };
 
-#if !NATIVEAOT // TODO
                 _committedCounter ??= new PollingCounter("gc-committed", this, () => ((double)GC.GetGCMemoryInfo().TotalCommittedBytes / 1_000_000)) { DisplayName = "GC Committed Bytes", DisplayUnits = "MB" };
+#if NATIVEAOT
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:136]: We should add the methods with the matching signatures instead of the ifdefs.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Assembly.NativeAot.cs [hunk 5]
```diff
@@ -35,6 +35,15 @@ public static Assembly Load(string assemblyString)
             return Load(name);
         }
 
+        // Performance metric to count the number of assemblies
+        // Caching since in NativeAOT, the number will be the same
+        private static uint? s_assemblyCount;
+        internal static uint GetAssemblyCount()
+        {
+            s_assemblyCount ??= (uint)AppDomain.CurrentDomain.GetAssemblies().Length;
+            return (uint)s_assemblyCount;
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Assembly.NativeAot.cs:44]:
```suggestion
        private static uint s_assemblyCount;
        internal static uint GetAssemblyCount()
        {
            if (s_assemblyCount == 0)
                s_assemblyCount = (uint)AppDomain.CurrentDomain.GetAssemblies().Length;
            return s_assemblyCount;
```
`uint?` is `Nullable<uint>`. Lock-free assignment of `Nullable<uint>` like this is not thread-safe.

Comment (agocke) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Assembly.NativeAot.cs:44]: Do we need a memory barrier here? Or is the idea that the threads can race and worst case you re-assign the same length?

Comment (jkotas) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Assembly.NativeAot.cs]: Yes, the idea is the threads can race to set this and last one wins.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Exception.NativeAot.cs [hunk 6]
```diff
@@ -112,6 +116,10 @@ private static void AppendExceptionStackFrame(object exceptionObj, IntPtr IP, in
                 bool isFirstFrame = (flags & (int)RhEHFrameType.RH_EH_FIRST_FRAME) != 0;
                 bool isFirstRethrowFrame = (flags & (int)RhEHFrameType.RH_EH_FIRST_RETHROW_FRAME) != 0;
 
+                // track count for metrics
+                if(isFirstFrame && !isFirstRethrowFrame)
+                    s_exceptionCount++;
```
Comment (agocke) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Exception.NativeAot.cs:121]: FYI this isn't thread-safe, you can lose increments. Not sure if it matters, @jkotas for feedback on how precise we want to be here, vs the cost of a memory barrier.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Exception.NativeAot.cs [hunk 7]
```diff
@@ -112,6 +116,10 @@ private static void AppendExceptionStackFrame(object exceptionObj, IntPtr IP, in
                 bool isFirstFrame = (flags & (int)RhEHFrameType.RH_EH_FIRST_FRAME) != 0;
                 bool isFirstRethrowFrame = (flags & (int)RhEHFrameType.RH_EH_FIRST_RETHROW_FRAME) != 0;
 
+                // track count for metrics
+                if(isFirstFrame && !isFirstRethrowFrame)
```
Comment (agocke) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Exception.NativeAot.cs:120]:
```suggestion
                if (isFirstFrame && !isFirstRethrowFrame)
```

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Exception.NativeAot.cs (hunk 6 above)
Comment (agocke) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/System/Exception.NativeAot.cs:121]: Also, open question on behavior for overflow. It seems like people might assume this value is monotonic.

Comment (jkotas) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/System/Exception.NativeAot.cs:121]: > Also, open question on behavior for overflow. It seems like people might assume this value is monotonic.

Mentioned in https://github.com/dotnet/runtime/issues/65989#issuecomment-1055018779 , but nothing was done about it.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/CompatibilitySuppressions.xml [hunk 8]
```diff
@@ -956,6 +956,38 @@
     <DiagnosticId>CP0001</DiagnosticId>
     <Target>T:System.Threading.LockHolder</Target>
   </Suppression>
+  <Suppression>
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/CompatibilitySuppressions.xml:959]: These can be revered. Reflection blocking was deleted from the product in #85810.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Exception.NativeAot.cs (hunk 6 above)
Comment (LakshanF) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/System/Exception.NativeAot.cs]: Created #85997 for the CoreCLR fix

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Assembly.NativeAot.cs [hunk 9]
```diff
@@ -35,6 +35,15 @@ public static Assembly Load(string assemblyString)
             return Load(name);
         }
 
+        // Performance metric to count the number of assemblies
+        // Caching since in NativeAOT, the number will be the same
+        private static uint? s_assemblyCount;
+        internal static uint GetAssemblyCount()
+        {
+            s_assemblyCount ??= (uint)AppDomain.CurrentDomain.GetAssemblies().Length;
```
Comment (agocke) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Assembly.NativeAot.cs:43]: What happened here? It looks like this was fixed, then reverted

**Review by agocke — CHANGES_REQUESTED** [review]

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Assembly.NativeAot.cs (hunk 9 above)
Comment (LakshanF) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Assembly.NativeAot.cs]: Sigh.. didn't sync correctly. Thanks for the catch!

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/GC.NativeAot.cs [hunk 10]
```diff
@@ -70,6 +70,16 @@ public static int GetGeneration(object obj)
             return RuntimeImports.RhGetGeneration(obj);
         }
 
+        public static int GetGenerationSize(int gen)
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/GC.NativeAot.cs:73]: Why are these public? It looks like they're just forcing unnecessary entries in CompatibilitySuppressions.xml - these are `internal` in CoreCLR's corelib.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Assembly.NativeAot.cs [hunk 11]
```diff
@@ -35,6 +35,16 @@ public static Assembly Load(string assemblyString)
             return Load(name);
         }
 
+        // Performance metric to count the number of assemblies
+        // Caching since in NativeAOT, the number will be the same
+        private static uint s_assemblyCount;
+        internal static uint GetAssemblyCount()
+        {
+            if (s_assemblyCount == 0)
+                s_assemblyCount = (uint)AppDomain.CurrentDomain.GetAssemblies().Length;
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Assembly.NativeAot.cs:44]: Since we're already in reflection implementation, no need to do the extra work AppDomain does:

```suggestion
                s_assemblyCount = (uint)Internal.Reflection.Core.Execution.ReflectionCoreExecution.ExecutionDomain.ReflectionDomainSetup.AssemblyBinder.GetLoadedAssemblies().Count;
```

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs [hunk 12]
```diff
@@ -46,12 +45,27 @@ public static class Keywords
...
         private IncrementingPollingCounter? _jitTimeCounter;
 
+#if NATIVEAOT
+        /// <summary>
+        /// If EventSource feature is enabled, RuntimeEventSource needs to be initialized
+        /// In CoreCLR, this is done via StartupHookProvider.CoreCLR.cs
+        /// </summary>
+#pragma warning disable CA2255
+        [ModuleInitializer]
+        internal static void NativeAOTtartupHook()
```
Comment (MichalStrehovsky) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:59]:
```suggestion
        internal static void NativeAOTStartupHook()
```

**Review by agocke — APPROVED** [review]
Lgtm

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs (hunk 12 above)
Comment (MichalStrehovsky) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs:59]: Or better yet - just put the `ModuleInitializer` on the `Initialize` method directly. The entire body of it is under `IsSupported` already and we're not getting anything out of this indirection.

**Review by MichalStrehovsky — CHANGES_REQUESTED** [review]

File: src/coreclr/nativeaot/System.Private.CoreLib/src/CompatibilitySuppressions.xml [hunk 13]
```diff
@@ -968,4 +968,4 @@
     <DiagnosticId>CP0016</DiagnosticId>
     <Target>M:System.Runtime.InteropServices.Marshal.GetObjectForNativeVariant``1(System.IntPtr)-&gt;T?:[T:System.Diagnostics.CodeAnalysis.MaybeNullAttribute]</Target>
   </Suppression>
-</Suppressions>
+</Suppressions>
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/CompatibilitySuppressions.xml:971]: Nit: Unnecessary whitespace change

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Exception.NativeAot.cs [hunk 14]
```diff
@@ -112,6 +117,10 @@ private static void AppendExceptionStackFrame(object exceptionObj, IntPtr IP, in
                 bool isFirstFrame = (flags & (int)RhEHFrameType.RH_EH_FIRST_FRAME) != 0;
                 bool isFirstRethrowFrame = (flags & (int)RhEHFrameType.RH_EH_FIRST_RETHROW_FRAME) != 0;
 
+                // track count for metrics
+                if (isFirstFrame && !isFirstRethrowFrame)
+                    Interlocked.Increment(ref s_exceptionCount);
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Exception.NativeAot.cs:122]: I'm going to assume this matches how CoreCLR counts these wrt rethrow or `ExceptionDispatchInfo.Throw`.

**Review by MichalStrehovsky — APPROVED** [review]
:shipit:
